test/handler/async.test.mo

```diff
@@ -71,85 +16,163 @@ func assert_state(x : (Nat, Nat, Nat)) {
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase fee while item still in queue (trigger did not run yet)
+await ledger.set_fee(6);
+ignore await* handler.updateFee();
+assert_state(0, 0, 0); // recalculation after fee update
+assert handler.journalLength() == inc(2); // #feeUpdated, #debited
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase deposit again
+await ledger.set_balance(7);
+assert (await* handler.notify(user1)) == ?(7, 1); // deposit = 7, credit = 1
+assert_state(7, 0, 1);
+assert handler.journalLength() == inc(2); // #newDeposit, #credited
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase fee while notify is underway (and item still in queue)
+// scenario 1: old_fee < previous = latest <= new_fee
+// this means no new deposit has happened (latest = previous)
+await ledger.lock_balance();
+let f1 = async { await* handler.notify(user1) }; // would return ?(0,1) at old fee
+await ledger.set_fee(10);
+ignore await* handler.updateFee();
+assert handler.journalLength() == inc(1); // #feeUpdated, not #debited because user1 is locked
+assert_state(7, 0, 1); // state still unchanged
+await ledger.release_balance(); // let notify return
+assert (await f1) == ?(0, 0); // deposit <= new fee
+assert_state(0, 0, 0); // state has changed
+assert handler.journalLength() == inc(1); // #debited
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase deposit again
+await ledger.set_balance(15);
+assert (await* handler.notify(user1)) == ?(15, 5); // deposit = 15, credit = 5
+assert_state(15, 0, 1);
+assert handler.journalLength() == inc(2); // #newDeposit, #credited
+print("tree lookups = " # debug_show handler.lookups());
+
+// increase fee while notify is underway (and item still in queue)
+// scenario 2: old_fee < previous <= new_fee < latest
+await ledger.set_balance(20);
+await ledger.lock_balance();
+let f2 = async { await* handler.notify(user1) }; // would return ?(5,10) at old fee
+await ledger.set_fee(15);
+ignore await* handler.updateFee();
+assert handler.journalLength() == inc(1); // #feeUpdated, not #debited because user1 is locked
+assert_state(15, 0, 1); // state still unchanged
+await ledger.release_balance(); // let notify return
+assert (await f2) == ?(20, 5); // credit = latest - new_fee
+assert_state(20, 0, 1); // state should have changed
+assert handler.journalLength() == inc(3); // #debited (recalculation), #newDeposit, #credited
+print("tree lookups = " # debug_show handler.lookups());
+
+// call multiple notify() simultaneously
+await ledger.lock_balance();
+let arr = [async { await* handler.notify(user1) }, async { await* handler.notify(user1) }, async { await* handler.notify(user1) }];
+await ledger.release_balance(); // let notify return
+assert (await arr[0]) == ?(0, 5); // first notify() should return state
+assert (await arr[1]) == null; // should return null
+assert (await arr[2]) == null; // should return null
```

Comment:
I would do this order:
```
assert (await arr[1]) == null; // should return null
assert (await arr[2]) == null; // should return null
await ledger.release_balance(); // let notify return
assert (await arr[0]) == ?(0, 5); // first notify() should return state
```
The null responses come immediately, they don't need the release_balance() call.